EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc

```diff
@@ -0,0 +1,176 @@
+      makeHCalTPGs(ctp7_phi, ctp7Data, res->getHcalDigis());
+      makeHFTPGs(ctp7_phi, ctp7Data, res->getHcalDigis());
+      makeRegions(ctp7_phi, ctp7Data, res->getRegions());
+
+      return true;
+   }
+
+   void CaloLayer1Unpacker::makeECalTPGs(uint32_t lPhi, UCTCTP7RawData& ctp7Data, EcalTrigPrimDigiCollection* ecalTPGs) {
+      UCTCTP7RawData::CaloType cType = UCTCTP7RawData::EBEE;
+      for(uint32_t iPhi = 0; iPhi < 4; iPhi++) { // Loop over all four phi divisions on card
+         int cPhi = - 1 + lPhi * 4 + iPhi; // Calorimeter phi index
+         if(cPhi == 0) cPhi = 72;
+         else if(cPhi == -1) cPhi = 71;
+         else if(cPhi < -1) {
+            LogError("CaloLayer1Unpacker") << "Major error in makeECalTPGs" << std::endl;
+            return;
+         }
+         for(int cEta =   -28; cEta <= 28; cEta++) { // Calorimeter Eta indices (HB/HE for now)
+            if(cEta != 0) { // Calorimeter eta = 0 is invalid
+               bool negativeEta = false;
+               if(cEta < 0) negativeEta = true;
+               uint32_t iEta = abs(cEta);
+               // This code is fragile! Note that towerDatum is packed as is done in EcalTriggerPrimitiveSample
+               // Bottom 8-bits are ET
+               // Then finegrain feature bit
+               // Then three bits have ttBits, which I have no clue about (not available on ECAL links so not set)
+               // Then there is a spare FG Veto bit, which is used for L1 spike detection (not available on ECAL links so not set)
+               // Top three bits seem to be unused. So, we steal those to set the tower masking, link masking and link status information 
+               // To decode these custom three bits use ((EcalTriggerPrimitiveSample::raw() >> 13) & 0x7)
+               uint32_t towerDatum = ctp7Data.getET(cType, negativeEta, iEta, iPhi);
+               if(ctp7Data.getFB(cType, negativeEta, iEta, iPhi)!=0) towerDatum |= 0x0100;
+               if(ctp7Data.isTowerMasked(cType, negativeEta, iEta, iPhi)) towerDatum |= 0x2000;
+               if(ctp7Data.isLinkMasked(cType, negativeEta, iEta, iPhi)) towerDatum |= 0x4000;
+               if(ctp7Data.isLinkMisaligned(cType, negativeEta, iEta, iPhi) ||
+                  ctp7Data.isLinkInError(cType, negativeEta, iEta, iPhi) ||
+                  ctp7Data.isLinkDown(cType, negativeEta, iEta, iPhi)) towerDatum |= 0x8000;
+               EcalTriggerPrimitiveSample sample(towerDatum); 
+               int zSide = cEta / ((int) iEta);
+               // As far as I can tell, the ECal unpacker only uses barrel and endcap IDs, never EcalTriggerTower
+               const EcalSubdetector ecalTriggerTower = (iEta > 17 ) ? EcalSubdetector::EcalEndcap : EcalSubdetector::EcalBarrel;
+               EcalTrigTowerDetId id(zSide, ecalTriggerTower, iEta, cPhi);
+               EcalTriggerPrimitiveDigi tpg(id);
+               tpg.setSize(1);
+               tpg.setSample(0, sample);
+               ecalTPGs->push_back(tpg);
```

Comment:
hi @rekovic - please reserve the size of ecalTPGs at the start of the double loop. (4*2*28 I think)